### PR TITLE
CDEF command line options

### PR DESCRIPTION
**Pull request by adaria**
| **Configuration file parameter** | **Command line** | **Range** | **Default** | **Description** |
| --- | --- | --- | --- | --- |
| **CDEFMode** | -cdef-mode | [0-5, -1 for default] | -1 | CDEF Mode, 0: OFF, 1-5: ON with 2,4,8,16,64 step refinement, -1: DEFAULT|

Issue #687

### CODE REVIEW

**Comment by adaria** [comment]
Investigating binary-identical-test failure

File: Source/Lib/Encoder/Codec/EbResourceCoordinationProcess.c [hunk 1]
```diff
@@ -186,6 +186,11 @@ EbErrorType signal_derivation_pre_analysis_oq(SequenceControlSet *     scs_ptr,
         scs_ptr->seq_header.enable_restoration =
             (uint8_t)scs_ptr->static_config.enable_restoration_filtering;
 
+    if (scs_ptr->static_config.enable_cdef == DEFAULT)
+        scs_ptr->seq_header.enable_cdef = 1;
+    else
+        scs_ptr->seq_header.enable_cdef = scs_ptr->static_config.enable_cdef;
+
     scs_ptr->cdf_mode = (pcs_ptr->enc_mode <= ENC_M6) ? 0 : 1;
```
Comment (adaria) [inline Source/Lib/Encoder/Codec/EbResourceCoordinationProcess.c]: The initialization added above was missing and causing the binary mismatch.
is `scs_ptr->cdf_mode` misnamed and should be `scs_ptr->cdef_mode` here?

File: Source/Lib/Encoder/Codec/EbPictureDecisionProcess.c [hunk 2]
```diff
@@ -986,17 +986,21 @@ EbErrorType signal_derivation_multi_processes_oq(
     // 3                                            8 step refinement
     // 4                                            16 step refinement
     // 5                                            64 step refinement
+    // TODO adaria: do I need to decouple enable_cdef and allow_intrabc here? at least signal to the user that they're coupled
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbPictureDecisionProcess.c]: they need to be coupled because they can not work together within the same stream

**Comment by hassount** [comment]
@adaria: can't we merge these together into 1 flag ? the reason why there are two of them is that one is set at the sequence header so that the encoder enables writing such in the bitstream then the decision is made per picture. so if the user would like it off, all we need to do is to shut the  sequence header flag off and so they could be all merged into 1 flag?

**Comment by adaria** [comment]
> @adaria: can't we merge these together into 1 flag ? the reason why there are two of them is that one is set at the sequence header so that the encoder enables writing such in the bitstream then the decision is made per picture. so if the user would like it off, all we need to do is to shut the sequence header flag off and so they could be all merged into 1 flag?

Yes, makes sense. I'll merge them into one flag with 0 for OFF and 1-5 for mode.

**Comment by adaria** [comment]
Will combine multiple on/off tools into a larger PR rather than adding them individually

File: Docs/Appendix-CDEF.md [hunk 3]
```diff
@@ -584,14 +584,14 @@ The search `in cdef_seg_search` and in `finish_cdef_search` for the
 filter strength is performed by considering a sub-interval of the filter
 strength index \[0,63\]. The subinterval is given by
 \[`cdef_ref_frame_strength` – `gi_step`, `cdef_ref_frame_strength` +
-`gi_step`\], where `cdf_ref_frame_strength` is the filter strength for
```
Comment (hassount) [inline Docs/Appendix-CDEF.md]: cdf is different from cdef in this case

**Pull request closed without merging**